rasterio/merge.py

```diff
@@ -396,73 +400,102 @@ def nullcontext(obj):
         else:
             chunks = [dout_window]
 
-        logger.debug("Chunks=%r", chunks)
+        def _window_frombounds(bounds, transform):
+            """Based on gdal_merge.py."""
+            minx, miny, maxx, maxy = bounds
+            xoff = int((minx - transform.c) / transform.a + 0.1)
+            yoff = int((maxy - transform.f) / transform.e + 0.1)
+            width = round((maxx - transform.c) / transform.a) - xoff
+            height = round((miny - transform.f) / transform.e) - yoff
+
+            if width < 1 or height < 1:
+                raise WindowError
+
+            return windows.Window(xoff, yoff, width, height)
+
+        def _intersect_bounds(bounds1, bounds2, transform):
+            """Based on gdal_merge.py."""
+            int_w = max(bounds1[0], bounds2[0])
+            int_e = min(bounds1[2], bounds2[2])
+
+            if int_w >= int_e:
+                raise ValueError
+
+            if transform.e < 0:
+                # north up
+                int_s = max(bounds1[1], bounds2[1])
+                int_n = min(bounds1[3], bounds2[3])
+                if int_s >= int_n:
+                    raise ValueError
+            else:
+                int_s = min(bounds1[1], bounds2[1])
+                int_n = max(bounds1[3], bounds2[3])
+                if int_n >= int_s:
+                    raise ValueError
+
+            return int_w, int_s, int_e, int_n
```

Comment:
Renamed and moved these functions since they're very specific to `merge()`.

Reply:
@sgillies Are there meaningful differences between `windows.from_bounds` and `_window_frombounds`? When writing these functions, I was trying to stay true to gdal_merge because at the time merge was have issues with rounding.

Reply:
There are. `windows.from_bounds()` avoids any rounding or aligning. The code in `gdal_merge.py` adds a fraction to offsets and lengths and then rounds down to get consistent values. In fact we already had that aligning behavior in the `Window` class.